.github/actions/hatch-setup/action.yml

```diff
@@ -25,37 +28,65 @@ runs:
       uses: pypa/hatch@install
 
     - name: Restore Hatch Directory
-      uses: actions/cache/restore@v4
-      id: cache-restore
+      uses: actions/cache@v4
+      id: cache
       with:
         path: |
           ${{ env.HATCH_CACHE_DIR }}
           ${{ env.HATCH_DATA_DIR }}
-        key: ${{ runner.os }}-hatch-${{ hashFiles('pyproject.toml') }}
+        key: ${{ runner.os }}-hatch-${{ hashFiles('pyproject.toml') }}-${{ hashFiles('requirements.txt') }}
 
-    - name: Install Dependencies
+    - name: Install Hatch Python
       shell: bash
-      if: steps.cache-restore.outputs.cache-hit != 'true'
+      if: steps.cache.outputs.cache-hit != 'true'
       run: |
         hatch python install ${{ inputs.min-python-version }} ${{ inputs.max-python-version }}
 
     - name: Install Dependencies
       shell: bash
-      if: steps.cache-restore.outputs.cache-hit != 'true'
+      if: steps.cache.outputs.cache-hit != 'true'
       run: |
         hatch env create
 
-    - name: Install Twine
+    - name: Test Matrix Present
+      id: test-matrix-present
       shell: bash
       run: |
-        hatch run python -m pip install twine
+        MATRIX_PRESENT=$(awk -v min="${{ inputs.min-python-version }}" -v max="${{ inputs.max-python-version }}" '
+            /\[tool\.hatch\.envs\.hatch-test\]/ { in_env=1; next }
+            in_env && /installer = "pip"/ { has_pip_installer=1; in_env=0 }
+            /\[\[tool\.hatch\.envs\.hatch-test\.matrix\]\]/ { in_matrix=1; next }
+            in_matrix && /python = \[/ {
+              if (index($0, min)) found_min=1
+              if (index($0, max)) found_max=1
+              in_matrix=0
+            }
+            END {
+              if (!has_pip_installer || !found_min || !found_max) {
+                print "false"
+                exit 0
+              }
+              print "true"
+              exit 0
+            }
+          ' pyproject.toml)
+        echo "matrix-present=$MATRIX_PRESENT" >> $GITHUB_OUTPUT
 
-    - name: Cache Hatch Directory
-      uses: actions/cache/save@v4
-      if: steps.cache-restore.outputs.cache-hit != 'true'
-      id: cache-hatch
-      with:
-        path: |
-          ${{ env.HATCH_CACHE_DIR }}
-          ${{ env.HATCH_DATA_DIR }}
-        key: ${{ runner.os }}-hatch-${{ hashFiles('pyproject.toml') }}
+    - name: Install Requirements.txt Dependencies for default python version
+      shell: bash
+      if: hashFiles('requirements.txt') != '' && steps.cache.outputs.cache-hit != 'true' && steps.test-matrix-present.outputs.matrix-present == 'false'
```

Comment:
Not following why `matrix-present` has to be false here.

Reply:
For matrix present we basicaly verify if the following settings are present in the pyproject.toml file.
```
[tool.hatch.envs.hatch-test]
installer = "pip"

# # Specify minimum and maximum Python versions to test
[[tool.hatch.envs.hatch-test.matrix]]
python = ["3.10", "3.12"]
```
Those settings allow hatch to use 2 separate environments with different versions of python and use preinstalled test dependencies. 
Without it we can only test reliably on one version of python.
So depending on if we have matrix setup in the settings or not. We need to call hatch differently.